include/deal.II/matrix_free/operators.h

```diff
@@ -1802,24 +1847,144 @@ namespace MatrixFreeOperators
     VectorType &diagonal_vector = this->diagonal_entries->get_vector();
     this->initialize_dof_vector(inverse_diagonal_vector);
     this->initialize_dof_vector(diagonal_vector);
-    inverse_diagonal_vector = Number(1.);
-    apply_add(diagonal_vector, inverse_diagonal_vector);
 
-    this->set_constrained_entries_to_one(diagonal_vector);
+    // Set up the action of the mass matrix in a way that's compatible with
+    // MatrixFreeTools::compute_diagonal:
+    auto diagonal_evaluation = [](auto &integrator) {
+      integrator.evaluate(EvaluationFlags::values);
+      for (unsigned int q = 0; q < integrator.n_q_points; ++q)
+        integrator.submit_value(integrator.get_value(q), q);
+      integrator.integrate(EvaluationFlags::values);
+    };
+
+    std::function<void(
+      FEEvaluation<
+        dim,
+        fe_degree,
+        n_q_points_1d,
+        n_components,
+        typename Base<dim, VectorType, VectorizedArrayType>::value_type,
+        VectorizedArrayType> &)>
+      diagonal_evaluation_f(diagonal_evaluation);
+
+    Assert(this->selected_rows.size() > 0, ExcInternalError());
+    for (unsigned int block_n = 0; block_n < this->selected_rows.size();
+         ++block_n)
+      MatrixFreeTools::compute_diagonal(*this->data,
+                                        BlockHelper::subblock(diagonal_vector,
+                                                              block_n),
+                                        diagonal_evaluation_f,
+                                        this->selected_rows[block_n]);
+
     inverse_diagonal_vector = diagonal_vector;
 
-    const unsigned int locally_owned_size =
-      inverse_diagonal_vector.locally_owned_size();
-    for (unsigned int i = 0; i < locally_owned_size; ++i)
-      inverse_diagonal_vector.local_element(i) =
-        Number(1.) / inverse_diagonal_vector.local_element(i);
+    for (unsigned int i = 0; i < inverse_diagonal_vector.locally_owned_size();
+         ++i)
+      if (std::abs(inverse_diagonal_vector.local_element(i)) >
+          std::sqrt(std::numeric_limits<Number>::epsilon()))
```

Comment:
Is this safe also when the vector entries are very small? As I see it length scales of 1e-6 in 3D would always fail here? I see where you want to go, but we would need some relative scaling here, e.g. `inverse_diagonal_vector.linfty_norm()`. I see that it adds some cost here but I think it should be worth it.

Reply:
You're right. In fact, I don't see any reason to check these values at all - the diagonal entries have to be positive or something has gone very wrong in the discretization.

I added an assertion and a call to `set_constrained_entries_to_one()`. I also added a check for division by zero in the lumped case. If the element doesn't support mass lumping we will still get negative entries in the vector (which the VectorTools code checks) but that will prevent floating point exceptions.